### DISCUSSION TITLE
Testing culture

### DISCUSSION THREAD

**Discussion by sberrevoets** [MobileNativeFoundation/discussions#9] [General]
Similar to https://github.com/MobileNativeFoundation/discussions/discussions/6 but with a different angle, I think many companies have had to define their own testing culture due to the absence of a strong testing culture in the community. On Apple platforms unit testing isn't instilled into developers from day one and there isn't great documentation available on how you can effectively test mobile apps, short of "here's how you use the frameworks".

It's a broad topic so the points below might be worth their own discussions entirely, but I figured this could kick us off. At Lyft we find these are big challenges for larger teams:

* Ensuring high test quality. We can look at other industries to help out, but mobile environments have unique challenges that don't necessarily apply elsewhere. How do you make sure everyone on the team understands how to write tests properly, the pitfalls in different environments (Rx, multithreading, UI, etc.), and _what_ to test? And how do you ensure people actually do write them?
* What architectural changes have teams made to build a more test-friendly app to lower the barrier of entry?
* Do "gamification" strategies help? Think of TDD, surfacing code coverage numbers, etc. For example, at Lyft we let people define their own minimum code coverage number based on what they think is reasonable for their module. If coverage drops below that number, CI will fail and they have to either decide to drop that number or write more tests. 9 out of 10 times we add more tests while still leaving the option open to just drop the number instead of writing `XCTAssertTrue(true)`-like tests.
* Developing the right guidelines around where tests are helpful and where they don't provide much value

**Comment by christian-bromann** [comment]
I also wonder if you all maintain separate UI tests for Android and iOS that test basically the same workflow but in different languages 🤔 I know that the industry moved away from writing one app with a framework like Cordova and release it on two platforms, mostly due to performance reasons I think, but I am curious if that might be still desired for testing.

**Reply by keith** [reply]
I posted a little about this here https://github.com/MobileNativeFoundation/discussions/discussions/6 but forgot to mention that we're writing the tests in the platform native languages. Lyft has tried that in the past a bit but never really got great results given the extra infrastructure involved. But I think if we could get that to work, it would mostly only improve the UI testing case, and I think many of @sberrevoets's questions would still apply for other cases

**Reply by AndreiVidrasco** [reply]
Have anyone tried using tools like https://github.com/Ahmed-Ali/Cucumberish and https://github.com/cucumber/cucumber-android
This way you can write `.feature` files, describing test scenario and be shared between platforms(making sure the behaviour match), but also in background writing the tests in the native languages

**Reply by keith** [reply]
I have seen some precedent for that on iOS over here https://github.com/bugsnag/bugsnag-cocoa/tree/master/features but I haven't heard of folks doing that in their apps. I'd definitely be interested to though!

**Reply by acadet** [reply]
The new kid around the block is [Waldo.io](https://www.waldo.io/) and could solve that issue 🤞 
It's a no code platform, I have not tried it personally though.

**Comment by codeman9** [comment]
I think that some of this has to do with how basic and unexpressive XCTest is. It doesn't really invite developers into desiring to write tests. I think that [RSpec](https://rspec.info/), for Ruby, on the other hand (even if you don't do TDD) uses language to write tests in a more inviting way. An analogous project that brings RSpec-like syntax to iOS development is [Quick](https://github.com/quick/quick) with [Nimble](https://github.com/quick/nimble). One of the best features of this framework is the `beforeEach` closures that let you systematically configure your objects as you go to allow sharing of setup beyond what XCTest provides.

Also, a good Dependency Injection regime is crucial to fostering good testing practices. Rather, when you start testing more things, you realize that you need to inject more dependencies so you have full control of the test scenarios. Now, I'm not talking about [Service Locator](https://stackoverflow.com/questions/22795459/is-servicelocator-an-anti-pattern), but actual [Dependency Injection](https://www.manning.com/books/dependency-injection-principles-practices-patterns).

Even though Apple doesn't instill the same sort of testing culture by default, that doesn't mean companies and teams can't adopt and make it part of their workflow. Just have to decide if it is worth it first and then do it, or not.

**Reply by khappucino** [reply]
+1 to this. The inability to really express your intent without RSpec is really a pain point.

**Reply by donnie-jp** [reply]
Definitely. Apple should adopt/recommend Quick/Nimble. 
The only decent argument I've seen against using it is that it's a 3rd party dependency so there's an element of risk involved.

**Reply by ArielDemarco** [reply]
We work in several modules, so part of our team uses Quick/Nimble, while others that are more comfortable with the native toolset code their tests in a BDD ("given-when-then") way, adding specific methods for each part of the test. It's really easy to read and kind of self-documented. It's pretty good also if you're doing TDD.
A simple example: 
```swift
func testXYZ() {
    givenAPrecondtion()
    whenInvokingSpecificMethod()
    thenSomethingHappens()
}
```

**Comment by PeqNP** [comment]
TDD provides rails for junior engineers and brings a general discipline to testing. It's not always necessary, but on large teams it probably makes sense to be somewhat strict about this.

Ensuring tests are written in a way that has team consensus. Ensuring the readability of tests is key, otherwise the value of the test is greatly diminished. For example, we have evolved to writing flat specs. Instead of nesting behaviors under different conditions (such as a feature flag is on or off), we write all of the conditions in the `describe`, delimited by a semi-colon. This will create test setup duplication, but keeping the setup as close to the `it`s as possible makes it easier to reference what is being done before the behavior is verified -- rather than walking up a tree of conditions that may or may not be grouped well. Of course these rules can be relaxed in some contexts. Especially, where clarity is better achieved by nesting.

I have also seen great advantages to pair programming. Knowledge transfer, getting a feature out "right" the first time, and team camaraderie alone pay for this process. What does this have to do with testing, you say? It solves one of the biggest problems of testing, which is that each pair ensures the other pair is writing a test (very important) and is using best test practices. For example, the description of the test can be immediately verified for clarity by their pair. That being said, Nordstrom went from exclusively pair programming to rarely doing it. While we do our best to maintain some amount of test best practices, it's becoming more difficult as the team grows larger.

Not sure if this is the right place for this gripe, but there is a definite lack of good tooling for test in iOS. Quick and Nimble gets you mostly there. I would like to see better testing grammar and facilities that ensure each test's state is encapsulated (to avoid test pollution) and clearly describes the purpose of the test. In other words, easy to read and write. Ideally, I would like to see a process that makes it dummy proof where 1. you provide a business rule that is being satisfied (without nesting state) 2. setup 3. subject action 4. verification. Most of the tools are too free-form. I feel wrong saying that as I appreciate flexibility, but it isn't necessary in most cases. For example, imagine a low-level API like OpenGL. Rarely does a developer interact with the API directly. It's usually through a framework's interface that abstracts much of the details away. In this way, I feel like XCTest fits this same pattern.

**Reply by sberrevoets** [reply]
Can you describe team size/structure a bit? What you say makes a lot of sense for teams of 10-15 people, but in our experience the cost of educating and onboarding new developers and making sure everyone is on the same page becomes a much bigger challenge when teams grow. The structure of the team (at Lyft, mobile developers are distributed to functional/feature teams, so there isn't a formal "iOS" or "mobile" team) also has a big impact on that.

All in all, the coordination efforts start to become so big that introducing frameworks like Quick and Nimble would take up so much time (not just once, but continuously) that you could seriously wonder if you're ever going to win that time back from having to fix fewer bugs.

**Reply by codeman9** [reply]
I worked with @PeqNP at Nordstrom and we definitely felt those same concerns when we started. Team size and structure varied quite a bit as time went on but we never wavered from our key values: 100% pairing and TDD. We went from a single iOS team/function, to squads, to feature teams, and even had other backend service teams start writing iOS code. The stipulation was that we would help them ramp up with our 100% pairing approach for a few weeks to teach them how to do things and then they went off on their own. Eventually, they became advocates for us to help educate others that wanted to submit iOS code and we just became approvers on their PRs for a time. In the end, I think that coming to work and doing what we were doing became a joy for almost everyone involved. This greatly overcame any concerns about winning time back. TDD is its own gamification of sorts. There is instant feedback and feelings of gratification when you or your pair writes a test that you make pass. Do that 100 times a day and you have a pair of happy developers.

**Reply by PeqNP** [reply]
Yes. What @codeman9 said.

I would advise against changing your testing framework. Especially for any reasonably sized project. There is plenty you can do with your existing framework that could help alleviate any friction you are experiencing. Such as how you logically group tests, test patterns to remove flakiness, and using single responsibility, etc. Rewriting tests would have to have organizational support behind it (similar to an app rewrite). I was simply lamenting the tools that are available -- especially as it relates to large teams. My apologies, as that wasn't clear. I don't want to tell anyone what tools to use. Too subjective and not helpful. I'm more interested in the general patterns as it relates to testing.

A slight tangent, and some context for the above:
Regarding single responsibility, often times I view the complexity of your test as a smell to the complexity of the subject under test. If your tests are large, complex, and hard to read, there's a good chance the API or the implementation is all of those things too. This is an opportunity to simplify the API and break out functional parts of the subject under test into separate classes/funcs/etc. using single responsibility.

Another method I use for testing is dependency inversion. As a consumer _how do *I* want to interact with my dependencies_? I think some refer to this as "protocol oriented programming". However you may refer to it, this pattern usually helps illuminate interfaces that are easier to consume... as the consumer is defining the interface.

To tie this altogether, these principles have helped me to write more clear tests and to help define interfaces (one of the most difficult aspects of programming). We try to instill these principles into our testing/development culture. I hope you find it useful.

**Reply by PeqNP** [reply]
I completely forgot about dependency injection too. This is such an important tool for testing. If you're not doing it, you really should. Try to avoid singletons and instead use an injection framework that manages the dependency graph for you (where a class is instantiated only once by the framework -- if needed). I've found testing to be nearly impossible for some classes that were singleton. There is too much opportunity for test pollution.

**Reply by codeman9** [reply]
Oh yeah, we managed to scale our team size from around 10 to near 60-70 or so dedicated iOS developers in Seattle and Kiev.

**Comment by ariel-demarco-peya** [comment]
In the teams I've participated in, what I've found the best tool to promote a testing culture is to show the value of it and the practices and patterns that are associated with testing.

Gamification always helps, but only if it's non-blocking. If that's the case you'll generate the opposite effect: people will tend to work for coverage, and not to really test their code. 

That said, what I think causes the problem in my experience is that many times we discuss and give a lot of importance to the overall Presentation Architecture (aka. MVC / MVVM / VIPER / RIB / etc) and try to follow an architecture pattern through all the app rather than learning the objectives that those architectures enclose and pursue. That's why, whatever doesn't fit into those patterns (or that has to be broadly shared), goes to a class (static or singleton) that tends to become untestable. Those objects themselves, cannot be injected as a whole (only behind an interface that hides the rusty implementation). At the same time, the mobile community focuses so much on these architectural patterns that they don't pay attention to what causes their code to be difficult to read/maintain/test in the first place, which is they generally don't use design patterns (Dependency Injection, Adapters, Facade, Factories, CoR, Strategies, etc) to solve problems or they don't pay attention to the design of their objects (following design principles like SOLID, Kiss, Dry, Slap, etc).

All of this happens as we tend to believe that the most difficult part of the frontend labor is coding the UI, but as time goes by, we realize it's not. The core of many of the apps is their model, not the UI. This can be seen when a UI rebranding is made: UI changes, but the app functionality is the same. That's why one of the approaches I've taken is to promote a more model-centric approach of structuring the code (Domain Driven Design or Interaction Driven Design). Those sets of tools focus on building blocks that follow certain guidelines (already established by the community) that are easier to test and follow.

At the same time, doing a model-centric encourages you to encapsulate business logic aside from UI and Infrastructure/peripherals. In that way, this approach will adapt whether you're using MVC, MVVM, VIPER (that, for me, should be feature/flow bounded and not for the whole app) or any other, while keeping the business (the important part of your app) fully tested.

The good thing about following this approach is that it tends to reduce the gap with other platforms (iOS/Android/Backend), as we go to a common ground that is business, where all of them can collaborate easily. 

As many of the developers I had to work with are not aware of these approaches, one of the things we do as a team is Coding Dojos / Katas that focus on design and testing. Also, for entry levels, we do hands-on workshops and we record those sessions so anyone is able to learn/follow them whenever they want. We do it regularly in teams (multidisciplinary), and also for newcomers. This allows everyone to learn the concepts in a controlled environment while looking at how things work in a real use case/exercise. Any feature/flow that's created will probably be done following the culture of well-crafted design where testing is a part of it.

**Reply by aaroncrespo** [reply]
My favorite exercise here is taking a bug report, and then replicating the bug it in the test suite. When the bug fix is ready your new test should now break verifying the fix. When you update the test for the fix this regression test now should prevent this bug from ever occurring again. 

The swift project is full of these types of tests.

**Reply by ArielDemarco** [reply]
That's actually great. When you reach that point the team clearly understands that testing is a way for you to work less, and not for you to work more.

**Comment by andreaantonioni** [comment]
Testing culture is a very interesting and hard topic in a team.

At Just Eat, after some "soft" solutions we tried (such as putting a checkbox in the PR description to check tests) we took a more "hard" one. Now we check the code coverage of each PR and the code added need to have 50% of unit test coverage otherwise the PR can't be merged.

This actually made a huge difference, it pushed engineers to start caring more about tests and how to write testable code.

**Reply by willwfsp** [reply]
Here at iFood, we started blocking merges on PRs that have no changes on the test targets. By doing it, our code coverage has increased from 10% to 70%. Although it took a little bit longer to achieve this number (2 years-ish), the project grew up absurdly fast.

**Reply by sberrevoets** [reply]
In either situation, how do you deal with PRs that only touch UI code/inherently untestable code, only move code around or are otherwise mechanical, or low quality tests to make sure the PR is mergeable?

**Reply by andreaantonioni** [reply]
That's a great question @sberrevoets, those were the reason I was strongly against this approach. But after some time, I have to admit that it's working great for us.

- UI code/untestable code: we started writing unit tests also for the UI where makes sense. For untestable code, it's quite rare we have it because with this enforcing engineers now think more about how to structure the code in order to make it testable. For rare exceptions, we ask a Staff engineer to review the PR if they're fine with the coverage then we merge it.
- Move code around: this probably means that you're moving code that hasn't tested in the past (before this enforcement), so we require you to add new tests.
- Low-quality tests: this is the point. We don't have a clear solution for this, an engineer could write low-quality tests to reach the targeted coverage and if none of the PR reviewers catches them, then it's done. For this, I can say that having 50% as a target gives enough flexibility to avoid adding unless tests. Also, we're organized in a similar way as Lyft so engineers are responsible for their features in their team, so if there's a bug in production they are appointed to fix it asap. This is a good motivation for engineers to make sure tests cover all-important parts of their features.

**Reply by codeman9** [reply]
With regard to UI code, I don't think it is inherently untestable. I've used [KIF](https://github.com/kif-framework/KIF) in the past and this is really great for putting a view or view controller on screen in the simulator and tapping buttons, verifying labels are visible, etc. It doesn't enforce widget placement on the screen, but there are other methods for that.  If `UIView`s are [Passive](https://martinfowler.com/eaaDev/PassiveScreen.html), then there isn't much to test anyway and if that is the case, then the tests become easy and straightforward. `UIView` tests end up looking like, "Put view on screen" then some sort of: "When a button is pressed, a delegate receives a message", "When items are on the screen, they are styled in this way", "When this button is on the screen, it is disabled until this other information is filled out", etc. We found it mainly important to test that our styling was applied, that our view delegates/helper objects were receiving the correct messages at the correct times, and that elements were hidden until a message was received from the view controller.

**Reply by ArielDemarco** [reply]
I think that if you apply some architectural pattern that abstracts `View` or `ViewControllers` so they're limited to just rendering logic, Snapshot tests are quite a good solution, and there're good libraries out there (like [iOSSnapshotTestCase](https://github.com/uber/ios-snapshot-test-case) or [SwiftSnapshotTesting](https://github.com/pointfreeco/swift-snapshot-testing).

Once you've tweaked some common issues (like dealing with render cycles in tables/collections or flaky tests on Xcode Migrations) they become a really useful tool for the team.

Also, those tests are really easy to make and they raise the coverage, so snapshot testing might be a good candidate if you're having trouble testing UI.

**Reply by willwfsp** [reply]
@sberrevoets Nice question. We do similar to the other folks' suggestion

- **UI code/untestable code:**  It's rare. We abstract the View from controllers and test it with snapshot testing as @ArielDemarco does. So every time we have a UI change, we have a test case that is supposed to be changed

- **Move code around:** We have a way to signalize the PR as TRIVIAL so the CI don't run the test job

- **Low-quality tests:** Here is where the test culture has more strength. The Reviewers reviews every single test case. We have a test case code style and some patterns that should be followed, and this makes the review easier. Also, we have Code Owners that have a full domain of some piece of code, so they suggest the best places that would be better to be covered. I think this code-owner methodology works very well because if some code is under your responsibility, you may review it better.

The most common untestable code changes are:
- Move code around without changing the component's Design/Architecture/Structure/API
- Some project configurations
- Performance improvements (but you can unit test it)

**Reply by MohabTarek** [reply]
Which tools you use to limit checks for unit test coverage on PR level?

**Reply by keith** [reply]
For our coverage case we parse the llvm-cov json output for the overall coverage % and diff that against a number the developer checks in

**Comment by freak4pc** [comment]
We're personally not huge fans of TDD. It's fun when learning how to develop but our team doesn't find it scalable or enjoyable to plan a solid test suite this way. We prefer understanding the behaviors we want to test and mainly consist of a mix of:

* Behavior-Driver Quick/Nimble/Cucumber-styled tests for Logic/ViewModel behavior
* Snapshot Tests for "UI Coverage" and also higher velocity for design reviews
* Black-box automation tests with Appium

**Comment by rudro** [comment]
We built our [RIB](https://github.com/uber/RIBs) architecture with testing in mind. It's one of a bigger reasons why we were ok with the large number of pieces we ended up with. The idea is that the business logic, routing logic, data presentation logic, can all be easily unit tested while limiting the amount of code that needs expensive and brittle UI testing. We do actively use [snapshot tests](https://github.com/uber/ios-snapshot-test-case/) to cover the views. We also use [dependency injection](https://github.com/uber/needle) everywhere to ensure testability.

We also use end-to-end (UI) tests for the full flow, but as expected, these are slow, somewhat brittle, and limited in number.

Finally, as you might expect, we do see requests for something that tests multiple RIBs (i.e. bigger than our typical unit tests) but are not full blown XCUITests.

**Comment by dinorahto** [comment]
I love this topic!
Cause I work in a company without QA (dedicated team), we are a huge startup here in Mexico, we are over 140 engineers with multiple codebases, multiple products and we don't even have outsourcing of anything and the vision of the CTO is never have a QA dedicated team. 

Databases, Data Science, Mining, Mobile, Back, API everything is testing by the developers that build it! and they have to cause that's the culture we have handle over the years, we believe that our code must be testable as much as functional in every layer possible! 
The Integration was in his moment a real and difficult topic, but we cross the chasm having the idea that we will grow as a startup if we do this in the long run!

**Comment by Julioevm** [comment]
Here at Xing we are trying to shift a bit our QA Strategy to be more unified across teams (each team was more or less deciding on the details of how they approached QA) and trying to make a whole team quality approach. 

Despite having a good coverage of test automation, we had an intensive manual process before releases, which we are trying to get rid of by having quality taken care of from the beginning of each feature implemented, building confidence to release without this regression.
 
Some teams had already taken into this mentality, others are probably going to be harder to shift, but the idea is that the whole platforms take this direction to support any issue teams might have with the transition.

We're also trying to remove the gatekeeper role from our testers: We think that this can give the impression that quality rests on the shoulders of testers, but now any dev can give the green light to merge a PR, so the responsibility is shared among the team.

I hope we can share positive results in the future!

**Reply by JoeSSS** [reply]
one of the ideas - removing dedicated QA columns in the dev process + always create documentation as a base for everything testing related to each feature.
